core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -307,9 +330,26 @@ public boolean exists(Settings settings) {
      * instead.
      */
     public T get(Settings settings) {
+        return get(settings, true);
+    }
+
+    private T get(Settings settings, boolean validate) {
         String value = getRaw(settings);
         try {
-            return parser.apply(value);
+            T parsed = parser.apply(value);
+            if (validate) {
+                final Iterator<Setting<T>> it = validator.settings();
+                final Map<Setting<T>, T> map;
+                if (it.hasNext()) {
+                    map = new HashMap<>();
+                    final Setting<T> setting = it.next();
+                    map.put(setting, setting.get(settings, false)); // we have to disable validation or we will stack overflow
+                } else {
+                    map = Collections.emptyMap();
+                }
+                validator.validate(parsed, map);
```

Comment:
I hear you. There is still the alternative of let people use Setting#get for most things as I think this is just an issue with circular validation (i.e., two settings that validate each other). Where there is a circular validation, I think we typically can solve it by validating only on one of the keys (as always validate all keys). If things really get complicated people can always use Setting#getRaw and skip validation (at the price of doing parsing). Just an idea to simplify this which I wanted to hear your thoughts on. Totally up to you.